src/bridge/token_bridge.cairo

```diff
@@ -0,0 +1,777 @@
+                .send_deposit_message(
+                    token,
+                    amount,
+                    appchain_recipient,
+                    message,
+                    constants::HANDLE_DEPOSIT_WITH_MESSAGE_SELECTOR,
+                );
+
+            self
+                .emit_deposit_event(
+                    token,
+                    amount,
+                    appchain_recipient,
+                    message,
+                    constants::HANDLE_DEPOSIT_WITH_MESSAGE_SELECTOR,
+                    nonce
+                );
+
+            // Piggy-back the deposit tx to check and update the status of token bridge deployment.
+            self.check_deployment_status(token);
+        }
+        //
+        //     checks token deployment status.
+        //     relies on starknet clearing l1-l2 message upon successful completion of deployment.
+        //     processing: check the l1-l2 deployment message. set status to `active` if consumed.
+        //     if not consumed after the expected duration, it returns the status to unknown.
+        //
+        fn check_deployment_status(ref self: ContractState, token: ContractAddress) {
+            let settings = self.token_settings.read(token);
+            if (settings.token_status == TokenStatus::Pending) {
+                return;
+            }
+
+            let _msg_hash = settings.deployment_message_hash;
+        // DEP(piltover) : to uncomment once the interface of piltover changes
+        // if (self.messaging_contract.read().sn_to_appchain_messages(msg_hash) > 0) {
+        //     let new_settings = TokenSettings {
+        //         token_status: TokenStatus::Active, ..settings
+        //     };
+        //     self.token_settings.write(token, new_settings);
+        // } else if (get_block_timestamp() > settings.pending_deployment_expiration) {
+        //     let new_settings = TokenSettings {
+        //         token_status: TokenStatus::Unknown,
+        //         deployment_msg_hash: 0,
+        //         pending_deployment_expiration: 0,
+        //         max_total_balance: 0,
+        //         withdrawal_limit_applied: false
+        //     };
+        //     self.token_settings.write(token, new_settings);
+        // }
+
+        }
+
+
+        fn withdraw(
+            ref self: ContractState,
+            token: ContractAddress,
+            amount: u256,
+            recipient: ContractAddress
+        ) {
+            assert(recipient.is_non_zero(), Errors::INVALID_RECIPIENT);
+
+            self.consume_message(token, amount, recipient);
+            let settings = self.token_settings.read(token);
+            // TODO: Consume quota from here
+            // DEP(byteZorvin): Complete the withdrawal component in cairo 
+            if (settings.withdrawal_limit_applied) {}
+            let tokenDispatcher = IERC20Dispatcher { contract_address: token };
+            tokenDispatcher.transfer(recipient, amount);
```

Comment:
looks like they do some nice checks on the L1 (is it to avoid reentrancy attacks?) - https://github.com/starknet-io/starkgate-contracts/blob/5a10fd263d29cd032b7229691d043520edae0737/starkware/solidity/libraries/Transfers.sol#L39. should we use a similar pattern?

Reply:
Yes those were for reentrancy attacks. Thank for pointing them out.
I have added the reentrancy guard component from OZ now

Reply:
How do we decide for which functions this guard is needed?

Reply:
Ideally this should be added before and after there is an external contract call. Right now I have added wherever there is an external contract call + fund outflow

Reply:
let's add guard for deposit as well